### ISSUE TITLE
Vendor micropip and lazily load packaging

### ISSUE THREAD

**Issue by hoodmane** [pyodide/pyodide#3724]
## 🚀 Feature

We should vendor micropip into the stdlib and lazily load packaging.

### Motivation

Currently it is quite hard to ensure that the strategy used to install packages is efficient. One must pick between micropip (longer load time, more flexible) and loadPyodide. There isn't a "pick the best way automatically" API. Many downstream applications always load micropip just in case, so their loading experience is always slowed down. In many cases, they don't need micropip but they are copying example code and deciding is too confusing. Example: pyscript currently __always__ installs micropip.

With diligence, it's possible to optimize this, but the code will generally look the same:
```js
const needsMicropip = !packages.every(pkg => pkg in pyodide._api.repodata_packages);
```

### Pitch

Split micropip into a portion that can work without packaging and a portion that requires it. Only load packaging when it is determined to be necessary. Add an argument to `loadPackage` that opts into loading packages from pypyi and an argument to micropip that requests to only load from the lock file. Vendor a subset of micropip into the Pyodide stdlib. This way we can optimize load time for people.

The main downside of this is that it's a fair amount of work...

**Comment by rth** [comment]
Well yes, actually now that pypackaging no longer depends on pyparsing (though we forgot to remove it https://github.com/pyodide/pyodide/pull/3726), micropip + pypackaging is just 70kB brotli compressed

![image](https://user-images.githubusercontent.com/630936/229266701-3840a5ab-5a28-4fd0-9dc4-42cb8b40d3bd.png)

so we could indeed just accept that cost load it by default, given the advantages it provides.

I'm not sure about vendoring, because it's not a stdlib package and it would be weird to put it in the same folder as stdlib modules. The problem is not so much that it's a separate download, but that it happens very late after all the wasm compilation. The figure below doesn't have the units on the time scale, but there is a significant delay  for loading micropip if it happens after the end of `loadPyodide`:

![image](https://user-images.githubusercontent.com/630936/229266636-1638dffa-663e-47fc-b7cb-82709844bd15.png)

A good solution IMO would be to have, `loadPyodide({..., packages=[...], install_micropip=true})` then,
 - we could start loading micropip as soon as `repodata.json` is loaded, and it would registered as a regular package, rather than something vendored
 - it's still possible to disable it. For instance when loading a known set packages  initially that are all in `repodata.json`, and no packages later, it's not actually useful to load micropip/
 - `packages` in `repodata.json` can also start pre-loading earlier which will reduce the overall load time.
 
 What I'm less sure is how this should work if we add more features to micropip. For instance adding alternative registries https://github.com/pyodide/micropip/issues/7 is becoming quite critical. But then if you want say a package from a different registry rather than `repodata.json` loaded first (maybe because it's a different version), this would impact the above logic. Maybe we would also add a `micropip_options={..}` eventually, and change the behavior based on the parameters there. Anyway we can probably deal with this once it happens in micropip.

**Comment by ryanking13** [comment]
> I'm not sure about vendoring, because it's not a stdlib package and it would be weird to put it in the same folder as stdlib modules.

If we think that micropip is a critical component of Pyodide, then I think it is okay to put micropip into stdlib packages since we already do that for `pyodide-py`. Putting `micropip` into `python_stdlib.zip` would also reduce the number of network requests.

If we go that way, then I guess we need to embed packaging into micropip so we don't have to load packaging separately.

**Comment by rth** [comment]
>  put micropip into stdlib packages 

Another thing to consider people are building things on top of micropip and packaging. So they can  have packages that specify `micropip` or `packaging` as a dependency (e.g. `piplite` in jupyterlite should do that) so if we vendor it, it would be downloaded twice unless we do some hacks to make it act as a regular packages while it's not. IMO it would unnecessarily create some edge cases that we don't necessarily want to deal with.  While pre-loading packages is something we need to deal with at some point anyway, so we might as well do that for micropip.

> reduce the number of network requests.

I'm not sure we care about this for  200kB+ files when the overhead of a single request is relatively low. On the contrary, multiple files can be downloaded in parallel, which might be easier for CDN ( it needs to compress it on the fly, and keep in cache on edge notes with probably some size-dependent rules). 

So it's not such an obvious question whether creating larger bundles in our case is a good thing. There might be some size threshold for this, but I'm not sure where it would be.

**Comment by rth** [comment]
Also, we kind of unvendored micropip from this repo so users can install new releases of micropip, if they need to (e.g. for bug fixes)..While vendoring it back would undo this effort. So overall I'm not in favor of it.

**Comment by ryanking13** [comment]
> Another thing to consider people are building things on top of micropip and packaging. So they can have packages that specify micropip or packaging as a dependency

Yes, that's a good point. I didn't consider packages that use micropip as a dependency.

I'm going to spend some time improving micropip for the next release, so let's see if we can come up with a good solution.
